### DISCUSSION TITLE
[cmake] How to distinguish between msys2 and mingw-w64

### DISCUSSION THREAD

**Discussion by silverqx** [msys2/MSYS2-packages#4706] [General]
Is there any way how can I distinguish between these two, I need to target MSYS2 UCRT64 specifically but not mingw-w64. CMake doesn't support this, the `if(MINGW)` targets both.

This is the code I want to use in MSYS2 but need to exclude it in mingw-w64:

```cmake
target_link_options(${target} INTERFACE
    $<$<CONFIG:Debug,RelWithDebInfo>:
        LINKER:--dynamicbase,--high-entropy-va,--nxcompat
        LINKER:--default-image-base-high>
    # Use faster linker ( GNU ld linker doesn't work with the Clang;
    # for both GCC and Clang )
    # TODO use LINKER_TYPE target property when min. version will be CMake v3.29 silverqx
    -fuse-ld=lld
)
```

**Comment by Biswa96** [comment]
I think "if WIN32 and not MSVC" would be appropriate in this case.

If you want to link with msys runtime then "if MSYS" would be correct one.

**Reply by silverqx** [reply]
No way, I don't believe 😅 I'll try it and let know.

**Reply by silverqx** [reply]
[`if(MSYS)`](https://cmake.org/cmake/help/latest/variable/MSYS.html) is only defined with the MSYS Makefiles generator, I'm using Ninja.

**Reply by silverqx** [reply]
Also good to tell I need to target MSYS2 UCRT64 and I don't care about MSYS2 mingw64.

**Reply by silverqx** [reply]
As I expected this doesn't work:

```cmake
if(MINGW)
    if(WIN32 AND NOT MSVC)
        message("||-- MSYS2 UCRT64")
    else()
        message("||-- mingw-w64")
    endif()
endif()
```

**Reply by silverqx** [reply]
Now I also tried this to be sure and doesn't work with Ninja:

```cmake

if(MINGW)
    if(MSYS)
        message("||-- MSYS2 UCRT64")
    else()
        message("||-- mingw-w64")
    endif()
endif()
```

**Reply by MehdiChinoune** [reply]
MINGW is not defined in MSYS
But MSYS is defined when using 'MSYS Makefiles' generator.

**Reply by MehdiChinoune** [reply]
CMake doesn't distinguish between UCRT (runtime of UCRT64,CLANG64...) and MSVCRT (runtime of MINGW64 and MINGW32).
You could distinguish between them using try_run CMake function and define variable according to its results.

**Reply by silverqx** [reply]
> MINGW is not defined in MSYS

This can be something, if it's true then I check it using eg. `check_cxx_source_compiles()` and decide based on the result.
